File: PMA.Api/Services/AiAssistantService.cs
using System.Text;
using PMA.Api.Interfaces;

namespace PMA.Api.Services;

public class AiAssistantService : IAiAssistantService
{
    public IReadOnlyList<string> SuggestRendezVousMotifs(string? patientDisplayName, DateTime? dateHeure, string? currentMotif)
    {
        var label = string.IsNullOrWhiteSpace(patientDisplayName) ? "patient" : patientDisplayName.Trim();
        var partOfDay = dateHeure.HasValue && dateHeure.Value.Hour >= 14 ? "suivi de l'apres-midi" : "suivi du matin";
        var seed = string.IsNullOrWhiteSpace(currentMotif) ? "consultation PMA" : currentMotif.Trim();

        return
        [
            $"{seed} - verification des examens et orientation",
            $"Entretien administratif avec {label} ({partOfDay})",
            "Point d'avancement du dossier et preparation des prochaines etapes"
        ];
    }

    public string ReformulateAdministrativeNote(string note)
    {
        var clean = NormalizeText(note);
        if (clean.Length == 0) return "";

        if (!clean.EndsWith(".", StringComparison.Ordinal))
            clean += ".";

        return $"Note administrative reformulee: {clean}";
    }

    public string SummarizePatientTimeline(PatientTimelineContext context)
    {
        var sb = new StringBuilder();
        var patient = string.IsNullOrWhiteSpace(context.PatientDisplayName) ? "Patient" : context.PatientDisplayName.Trim();
        var dossier = string.IsNullOrWhiteSpace(context.DossierType) ? "non precise" : context.DossierType.Trim();

        sb.Append($"{patient} - dossier {dossier}. ");

        if (context.RendezVous.Count == 0)
        {
            sb.Append("Aucun rendez-vous enregistre. ");
        }
        else
        {
            var ordered = context.RendezVous.OrderBy(r => r.DateHeure).ToList();
            var next = ordered.FirstOrDefault(r => r.DateHeure >= DateTime.Now);
            sb.Append($"{ordered.Count} rendez-vous au total");
            if (next is not null)
                sb.Append($", prochain le {next.DateHeure:dd/MM/yyyy HH:mm}");
            sb.Append(". ");
        }

        if (context.Cycles.Count > 0)
        {
            var actifs = context.Cycles.Count(c =>
                !string.Equals(c.StatutCycle, "termine", StringComparison.OrdinalIgnoreCase));
            sb.Append($"{context.Cycles.Count} cycle(s) PMA, dont {actifs} actif(s). ");
        }

        if (context.Actes.Count > 0)
        {
            var aFaire = context.Actes.Count(a =>
                string.Equals(a.StatutRealisation, "a_faire", StringComparison.OrdinalIgnoreCase));
            sb.Append($"{context.Actes.Count} acte(s) prescrit(s), {aFaire} encore a realiser.");
        }

        return sb.ToString().Trim();
    }

    private static string NormalizeText(string value)
    {
        var tokens = value
            .Replace("\r", " ", StringComparison.Ordinal)
            .Replace("\n", " ", StringComparison.Ordinal)
            .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        if (tokens.Length == 0) return string.Empty;

        var joined = string.Join(" ", tokens);
        if (joined.Length == 1) return joined.ToUpperInvariant();
        return char.ToUpperInvariant(joined[0]) + joined[1..];
    }
}

File: pma-front/src/app/core/services/ai-assistant.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ActePma, CyclePma, RendezVous } from '../models';

interface SuggestRdvMotifResponse {
  suggestions: string[];
}

interface ReformulateNoteResponse {
  reformulatedNote: string;
}

interface SummarizePatientResponse {
  summary: string;
}

@Injectable({ providedIn: 'root' })
export class AiAssistantService {
  private readonly apiUrl = `${environment.apiUrl}/ai-assistant`;

  constructor(private http: HttpClient) {}

  suggestRdvMotif(payload: {
    patientId?: number;
    patientDisplayName?: string;
    dateHeure?: string;
    currentMotif?: string;
  }): Observable<SuggestRdvMotifResponse> {
    return this.http.post<SuggestRdvMotifResponse>(`${this.apiUrl}/suggest-rdv-motif`, payload).pipe(
      catchError(() => {
        const seed = (payload.currentMotif || '').trim() || 'Consultation PMA';
        const suggestions = [
          `${seed} - verification du dossier`,
          'Entretien administratif et planification des etapes',
          'Suivi patient et coordination avec le laboratoire'
        ];
        return of({ suggestions });
      })
    );
  }

  reformulateNote(note: string): Observable<ReformulateNoteResponse> {
    return this.http.post<ReformulateNoteResponse>(`${this.apiUrl}/reformulate-note`, { note }).pipe(
      catchError(() => {
        const normalized = this.normalizeText(note);
        const reformulatedNote = normalized
          ? `Note administrative reformulee: ${normalized.endsWith('.') ? normalized : `${normalized}.`}`
          : '';
        return of({ reformulatedNote });
      })
    );
  }

  summarizePatient(payload: {
    patientDisplayName?: string;
    dossierType?: string;
    rendezVous: Pick<RendezVous, 'dateHeure' | 'motif' | 'statut'>[];
    cycles: Pick<CyclePma, 'dateDebut' | 'phase' | 'etapeCourante' | 'statutCycle'>[];
    actes: Pick<ActePma, 'libelle' | 'typeActe' | 'statutRealisation'>[];
  }): Observable<SummarizePatientResponse> {
    return this.http.post<SummarizePatientResponse>(`${this.apiUrl}/summarize-patient`, payload).pipe(
      map((res) => ({ summary: (res.summary || '').trim() })),
      catchError(() => {
        const summary = this.buildLocalSummary(payload);
        return of({ summary });
      })
    );
  }

  private buildLocalSummary(payload: {
    patientDisplayName?: string;
    dossierType?: string;
    rendezVous: Pick<RendezVous, 'dateHeure' | 'motif' | 'statut'>[];
    cycles: Pick<CyclePma, 'dateDebut' | 'phase' | 'etapeCourante' | 'statutCycle'>[];
    actes: Pick<ActePma, 'libelle' | 'typeActe' | 'statutRealisation'>[];
  }): string {
    const patient = (payload.patientDisplayName || 'Patient').trim();
    const dossier = (payload.dossierType || 'non precise').trim();
    const rdvCount = payload.rendezVous.length;
    const cyclesCount = payload.cycles.length;
    const actesCount = payload.actes.length;
    const actifs = payload.cycles.filter((c) => (c.statutCycle || '').toLowerCase() !== 'termine').length;
    const aFaire = payload.actes.filter((a) => (a.statutRealisation || '').toLowerCase() === 'a_faire').length;

    return `${patient} - dossier ${dossier}. ${rdvCount} rendez-vous, ${cyclesCount} cycle(s) dont ${actifs} actif(s), ${actesCount} acte(s) prescrit(s) dont ${aFaire} a realiser.`;
  }

  private normalizeText(value: string): string {
    const tokens = value
      .replace(/\r/g, ' ')
      .replace(/\n/g, ' ')
      .split(' ')
      .map((x) => x.trim())
      .filter(Boolean);
    if (tokens.length === 0) return '';
    const joined = tokens.join(' ');
    return joined[0].toUpperCase() + joined.slice(1);
  }
}
